packages/fx-core/src/core/middleware/projectMigrator.ts

```diff
@@ -280,6 +285,44 @@ async function migrateToArmAndMultiEnv(ctx: CoreHookContext): Promise<void> {
   await postMigration(projectPath, ctx, inputs);
 }
 
+async function preCheckKeyFiles(projectPath: string, ctx: CoreHookContext): Promise<void> {
+  const core = ctx.self as FxCore;
+  const fx = path.join(projectPath, `.${ConfigFolderName}`);
+  const manifestPath = (await fs.pathExists(path.join(fx, AppPackageFolderName, REMOTE_MANIFEST)))
+    ? path.join(fx, AppPackageFolderName, REMOTE_MANIFEST)
+    : path.join(projectPath, AppPackageFolderName, REMOTE_MANIFEST);
+
+  await preReadJsonFile(path.join(fx, "env.default.json"), core);
+  await preReadJsonFile(path.join(fx, "settings.json"), core);
+  await preReadJsonFile(manifestPath, core);
+}
+
+async function preReadJsonFile(path: string, core: FxCore): Promise<void> {
+  try {
+    await fs.readJson(path);
+  } catch (err) {
+    core.tools.logProvider.error(
+      `'${path}' is not exist or not in json format. Please fix it and try again by running command (Teams: Upgrade project).`
+    );
+    core.tools.logProvider.warning(`Read this wiki(${learnMoreLink}) for the FAQ.`);
+
+    core.tools.ui
+      .showMessage(
+        "info",
+        util.format(getStrings().solution.MigrationToArmAndMultiEnvPreCheckErrorMessage, path),
+        false,
+        learnMoreText
+      )
+      .then((result) => {
+        const userSelected = result.isOk() ? result.value : undefined;
+        if (userSelected === learnMoreText) {
+          core.tools.ui!.openUrl(manualUpgradeLink);
+        }
+      });
+    throw NotJsonError(err);
```

Comment:
nit: how about rename it to `ReadJsonFileError` since there're maybe two kinds of error here:
1. File not found.
2. invalid JSON

`NotJsonError` is not that accurate from readability perspective.